### DISCUSSION TITLE
EC public key serialization changes in openssl 3

### DISCUSSION THREAD

**Discussion by matthauck** [openssl/openssl#22086] [Q&A] [answered]
We are looking to move to openssl 3.0 from openssl 1.0.2, and noticed a change in public key serialization for some old public keys we have. (Note: these keys were not originally generated by openssl.)

Our application uses the hash of the public key's DER serialization to identify public keys. And it is rather crucial that keys get identified the same across this upgrade.

Here is an example public key that is serialized differently on openssl1 vs. openssl3

```
$ ~/usr/openssl-3.0.10/bin/openssl ec -in Components/Testing/DataFiles/TestPublicKey.pub  -inform der -pubin
read EC key
writing EC key
-----BEGIN PUBLIC KEY-----
MIICRjCCAbkGByqGSM49AgEwggGsAgEBME0GByqGSM49AQECQgH/////////////
////////////////////////////////////////////////////////////////
/////////zCBiARCAf//////////////////////////////////////////////
///////////////////////////////////////8BEIAUZU+uWGOHJofkpohoLaF
QO6i2nJbmbMV87i0iZGO8QnhVhk5Uex+k3sWUsC9O7G/BzVz34g9LDTx70Uf1GtQ
PwAEgYUEAMaFjga3BATpzZ4+y2YjlbRCnGSBOQU/tSH4KK9ga009uqFLXnfv51ko
/h3BJ6L/qN4zSLPBhWpCm/l+fjHC5b1mARg5KWp4mjvABFyKX7QsfRvZmPVESVeb
RGgXr70XJz5mLJfucple9CZAxVC5AT+tB2E1PHCGonLCQIi+lHaf0WZQAkIB////
///////////////////////////////////////6UYaHg78vlmt/zAFI9wml0Du1
ybiJnEeuu2+3HpE4ZAkCAQEDgYYABACKnYs6ZaOCVqd1Cte0ttmeq11sSxmrBEz+
rWQKBMcNK/l4kUYYewncFAUZzmrGckeS2+QXbY0nxFZ9dxeEHShV+gCt0Ka+7LI+
z6DPF9bfojirUDnn7wkAZb8ETk7DE8BruXnAal7Bcp2gQaWb/a3BxQuk+YsZ2xBu
glcNKvdAvNaF9Q==
-----END PUBLIC KEY-----
```

```
$ ~/usr/openssl-1.0.2zi/bin/openssl  ec -in Components/Testing/DataFiles/TestPublicKey.pub  -inform der -pubin
read EC key
writing EC key
-----BEGIN PUBLIC KEY-----
MIICRTCCAbgGByqGSM49AgEwggGrAgEBME0GByqGSM49AQECQgH/////////////
////////////////////////////////////////////////////////////////
/////////zCBhwRCAf//////////////////////////////////////////////
///////////////////////////////////////8BEFRlT65YY4cmh+SmiGgtoVA
7qLacluZsxXzuLSJkY7xCeFWGTlR7H6TexZSwL07sb8HNXPfiD0sNPHvRR/Ua1A/
AASBhQQAxoWOBrcEBOnNnj7LZiOVtEKcZIE5BT+1Ifgor2BrTT26oUted+/nWSj+
HcEnov+o3jNIs8GFakKb+X5+McLlvWYBGDkpaniaO8AEXIpftCx9G9mY9URJV5tE
aBevvRcnPmYsl+5ymV70JkDFULkBP60HYTU8cIaicsJAiL6Udp/RZlACQgH/////
//////////////////////////////////////pRhoeDvy+Wa3/MAUj3CaXQO7XJ
uImcR667b7cekThkCQIBAQOBhgAEAIqdizplo4JWp3UK17S22Z6rXWxLGasETP6t
ZAoExw0r+XiRRhh7CdwUBRnOasZyR5Lb5BdtjSfEVn13F4QdKFX6AK3Qpr7ssj7P
oM8X1t+iOKtQOefvCQBlvwROTsMTwGu5ecBqXsFynaBBpZv9rcHFC6T5ixnbEG6C
Vw0q90C81oX1
-----END PUBLIC KEY-----
```

**Comment by davidben** [comment, marked as answer]
This comes from https://github.com/openssl/openssl/pull/6314, which was made back in 2018. The new serialization is correct. The old one came from older OpenSSLs failing to represent field elements as fixed-width, but specification for this format is unambiguous.

(As an aside, this appears to be a specifiedCurve encoding of P-521. What it really should be using is the namedCurve form, which is both safer and less than 1/3 of the size. The specifiedCurve form is banned by [RFC 5480](https://www.rfc-editor.org/rfc/rfc5480#section-2.1.1), published in 2009. It has also been the cause of a few security bugs over the years, including CVE-2022-0778 and CVE-2020–0601. If you have something that is generating keys in this way, you may wish to re-examine it. You can get the namedCurve form by passing `-param_enc named_curve` to `openssl ec`.)

I think this bug can be closed.

**Comment by matthauck** [comment]
(Yeah. These old keys are indeed odd -- should have been named, but instead specified all the params. 🤷 We don't generate keys this way anymore.)

ok. thanks.
